components/reusable-components/tags-modal/TagsModal.tsx

```diff
@@ -0,0 +1,95 @@
+/* eslint-disable jsx-a11y/no-static-element-interactions */
+/* eslint-disable jsx-a11y/click-events-have-key-events */
+
+'use client';
+
+import React, { ChangeEvent, useState } from 'react';
+import styles from './TagsModal.module.scss';
+
+interface TagSearchProps {
+  existingTags: string[];
+  selectedTags: string[];
+  setSelectedTags: (items: string[]) => void;
+  onClose: () => void;
+}
+
+const TagsModal = ({ existingTags, selectedTags, setSelectedTags, onClose }: TagSearchProps) => {
+  const [searchTag, setSearchTag] = useState<string>('');
+  const [filteredTags, setFilteredTags] = useState<string[]>([]);
+
+  const handleSearchChange = (e: ChangeEvent<HTMLInputElement>) => {
+    const query = e.target.value;
+    setSearchTag(query);
+
+    if (query) {
+      const filtered = existingTags.filter((tag) =>
+        tag.toLowerCase().includes(query.toLowerCase())
+      );
+      setFilteredTags(filtered);
+    } else {
+      setFilteredTags([]);
+    }
+  };
+
+  const addTag = (tag: string) => {
+    if (!selectedTags.includes(tag)) {
+      setSelectedTags([...selectedTags, tag]);
+    }
+    setSearchTag('');
+    setFilteredTags([]);
+  };
+
+  const removeTag = (tag: string) => {
+    setSelectedTags(selectedTags.filter((item) => item !== tag));
+  };
```

Comment:
To avoid issues with stale state, we should use  the snapshot of the most recent state. something like:   setSelectedTags((prevTags) ....